Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoASA.yaml

```diff
@@ -0,0 +1,312 @@
+      | extend ASimMatchingIpAddr = case(
+                                        array_length(src_or_any) == 0 and array_length(dst_or_any) == 0, "-", // match not requested
+                                        (temp_isSrcMatch and temp_isDstMatch), "Both", // has to be checked before the individual 
+                                        temp_isSrcMatch, "SrcIpAddr",
+                                        temp_isDstMatch, "DstIpAddr",
+                                        "No match"
+                                    )
+      | where ASimMatchingIpAddr != "No match"
+      | where ((array_length(dvcaction) == 0) or SimplifiedDeviceAction has_any (dvcaction));
+      let all_106001_alike = allLogs
+      | where DeviceEventClassID in ("106001", "106006", "106015", "106016", "106021", "106022") 
+      | parse Message with * " interface " DstInterfaceName;
+      let all_106010_alike = allLogs
+      | where DeviceEventClassID in ("106010", "106014")
+      | parse Message with * " src "  SrcInterfaceName ":" * " dst " DstInterfaceName ":" *;
+      let all_106018 = allLogs
+      | where DeviceEventClassID == "106018"
+      | parse Message with * "list " NetworkRuleName:string " " *;
+      let all_106023 = allLogs
+      | where DeviceEventClassID == "106023" and not(Message contains "protocol 41") and isnotempty(SourceIP)
+      | parse Message with * " src "  SrcInterfaceName ":" * " dst " DstInterfaceName ":" * ' by access-group "' NetworkRuleName '" '  *;
+      let all_106023_unparsed = allLogs
+      | where DeviceEventClassID == "106023" and isempty(SourceIP) and not(Message contains "protocol 41")
+      | where Message has tostring(dstportnumber)
+      | extend parsedData = extract_all(@'%ASA-\d-\d+:\s*(\w+)\s+(\w+)\s+src\s+(\S+):(\d+.\d+.\d+.\d+)\/(\d+)\s*(\(\S+\)|\S+)*\s+dst\s+(\S+):(\d+.\d+.\d+.\d+)\/(\d+)\s*([(\S+\s+)]+|\S+)*\s*by\s+access-group\s*"(\S+)"',Message)[0]
+      | extend 
+              DeviceAction = coalesce(DeviceAction, tostring(parsedData[0])),
+              Protocol = coalesce(Protocol, tostring(parsedData[1])),
+              SrcInterfaceName = tostring(parsedData[2]),
+              SrcIpAddr = coalesce(SourceIP, tostring(parsedData[3])),
+              SrcPortNumber = coalesce(SourcePort, toint(parsedData[4])),
+              SrcUsername = replace_string(replace_string(tostring(parsedData[5]), ")", ""), "(", ""),
+              DstInterfaceName = tostring(parsedData[6]),
+              DstIpAddr = coalesce(DestinationIP, tostring(parsedData[7])),
+              DstPortNumber = coalesce(DestinationPort, toint(parsedData[8])),
+              DstUsername = replace_string(replace_string(tostring(parsedData[9]), ")", ""), "(", ""),
+              NetworkRuleName = tostring(parsedData[10])
+      | where (isnull(dstportnumber) or (DstPortNumber == dstportnumber))
+      | extend temp_isSrcMatch=has_any_ipv4_prefix(SrcIpAddr,src_or_any), 
```

Comment:
- You need to prefilter ealier in the all_106023_unparsed branch. True for all "heavy parsing" branches.
- You can use functions to reduce the repetition.